src/stan_math_backend/Transform_Mir.ml

```diff
@@ -497,9 +499,9 @@ and validate_dims_stmt stmt =
 
 let make_fill vident st loc =
   let rhs =
-    Expr.(
-      Helpers.internal_funapp FnNaN []
-      @@ Typed.Meta.create ~type_:UReal ~loc ~adlevel:DataOnly ())
+    Expr.
+      { Fixed.pattern= Var "DUMMY_VAR__"
```

Comment:
I think the optimizer might break when using a Var that hasn't been defined anywhere - I think those might need special consideration. How do you feel about trying to add the DUMMY_VAR__ declaration to the MIR before we pretty print it? Then we can remove all of the DUMMY_VAR__ string printing in Stan_math_code_gen and maybe have the optimizer work correctly without modifying it. May need to turn on dead code elimination by default but that would be a nice little win too, to avoid initializing the DUMMY_VAR__ even when it's never used.

An alternative is to modify the optimizer to add another special Var that it shouldn't worry about and see if that fixes it. Perhaps it has a bug around this behavior even after that's fixed.

Reply:
Great! Assigning `FnNaN()` to `DUMMY_VAR__` at the start of the block does fix it! (Declaration didn't matter)

However, when the optimizer sees `DUMMY_VAR__` fully it can and will undo all the work here. The point was after all to create a single NaN-valued `vari` and reuse it everywhere but the optimizer will happily deduce that `DUMMY_VAR__` is `DataOnly`.

I'm starting to think it's better to just not have any initialization in the MIR. Arrays without Eigen matrices are initialized by the backend without any hint in the MIR so why not Eigen types too.

Reply:
Just a couple notes in case this wasn't obvious.

If `stan::math::var` is declared but not defined, the nullary constructor sets its member `vari*` pointer to be undefined.  So if you try to extract a value or adjoint it'll cause problems.  But if you call `var(double x)` instead, then you get memory allocation for the `vari` and its value is set to `x`.  Therefore if initializing to a value, like 0 or even NaN, it's better to create a single `var` (equivalently `vari*`) and copy it everywhere, because there's no extra allocation on the autodiff stack.

Sorry if that was obvious.  With `std::vector`, it's efficient to initialize a size `n` standard vector to `x` using `std::vector<T> y(n, T(x));`